cpp/core/utils/ObjectStore.cc

```diff
@@ -37,23 +37,34 @@ gluten::ResourceMap<gluten::ObjectStore*>& gluten::ObjectStore::stores() {
 }
 
 gluten::ObjectStore::~ObjectStore() {
-  // destructing in reversed order (the last added object destructed first)
-  const std::lock_guard<std::mutex> lock(mtx_);
-  for (auto itr = aliveObjects_.rbegin(); itr != aliveObjects_.rend(); itr++) {
-    const ResourceHandle handle = (*itr).first;
-    const auto& info = (*itr).second;
-    const std::string_view typeName = info.typeName;
-    const size_t size = info.size;
-    VLOG(2) << "Unclosed object ["
-            << "Store ID: " << storeId_ << ", Resource handle ID: " << handle << ", TypeName: " << typeName
-            << ", Size: " << size
-            << "] is found when object store is closing. Gluten will"
-               " destroy it automatically but it's recommended to manually close"
-               " the object through the Java closing API after use,"
-               " to minimize peak memory pressure of the application.";
-    store_.erase(handle);
+  std::vector<std::shared_ptr<void>> objects;
+  {
+    // destructing in reversed order (the last added object destructed first)
+    const std::lock_guard<std::mutex> lock(mtx_);
+    for (auto itr = aliveObjects_.rbegin(); itr != aliveObjects_.rend(); itr++) {
+      const ResourceHandle handle = (*itr).first;
+      const auto& info = (*itr).second;
+      const std::string_view typeName = info.typeName;
+      const size_t size = info.size;
+      VLOG(2) << "Unclosed object ["
+              << "Store ID: " << storeId_ << ", Resource handle ID: " << handle << ", TypeName: " << typeName
+              << ", Size: " << size
+              << "] is found when object store is closing. Gluten will"
+                 " destroy it automatically but it's recommended to manually close"
+                 " the object through the Java closing API after use,"
+                 " to minimize peak memory pressure of the application.";
+      // transfer ownership of the object to the temporary vector
+      objects.push_back(store_.lookup(handle));
+      store_.erase(handle);
+    }
+    stores().erase(storeId_);
+  }
+  // destructing objects outside the lock to avoid deadlock
+  for (auto& obj : objects) {
+    if (obj) {
+      obj.reset();
```

Comment:
In the case your mentioned, `store_` may be accessed again when `obj.reset()` is called. However, IIUT, at this moment, the main object container `store_` is already emptied. Wouldn't there be a new `object not found` error being thrown, which is unexpected either?

Reply:
You are right but in this situation there will be no problem for ParquetFileWriter, which will ignore all exception during it's destructors as https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/writer/arrow/FileWriter.cpp#L646.

`In C++, destructors should generally not throw exceptions that are allowed to escape the destructor. This is a crucial guideline for writing robust and reliable C++ code`.

For other obj  besides `ParquetFileWriter`, if `object not found` being thrown we can fix it's destructors method, it's better to run in endless deadLock loop.

Reply:
Thank you for the explanation. I'll take a closer look tomorrow on this. You are right that at least we can avoid a dead lock by the change.

Reply:
@zjuwangg Would you think the following could be feasible (pseudo)? Thanks!

```cpp
for (;;) {
  lock_.lock();
  if (aliveObjects_.empty()) {
    break;
  }
  obj o = aliveObjects_.pop()
  lock_.unlock()
  o.release();
}
```